src/level_four/tap_cards/controller/delete_from_cards_controller.c

```diff
@@ -57,113 +57,190 @@
+ * STATIC VARIABLES
+ *****************************************************************************/
 
-    case TAP_CARD_FOUR_BACKEND:
-      _tap_card_backend(4);
-      break;
-    default:
-      break;
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static void delete_wallet_flash(uint8_t flash_wallet_index) {
+  ASSERT(SUCCESS_ == delete_wallet_share_from_sec_flash(flash_wallet_index));
+  ASSERT(SUCCESS_ == delete_wallet_from_flash(flash_wallet_index));
+
+  return;
+}
+
+static void handle_wallet_deleted_from_card(
+    card_delete_share_cfg_t *delete_config) {
+  uint8_t flash_wallet_index = 0xFF;
+
+  ASSERT(SUCCESS ==
+         get_index_by_name((const char *)delete_config->wallet->wallet_name,
+                           &flash_wallet_index));
+  ASSERT(SUCCESS == delete_from_kth_card_flash(flash_wallet_index,
+                                               delete_config->card_number));
+
+  if (4 == delete_config->card_number) {
+    delete_wallet_flash(flash_wallet_index);
   }
+
+  return;
 }
 
-// Card number is 1,2,3 or 4
-static void _tap_card_backend(uint8_t card_number) {
-  uint8_t flash_wallet_index;
-
-  if (get_index_by_name((const char *)wallet.wallet_name,
-                        &flash_wallet_index) != SUCCESS_)
-    return;
-  memcpy(tap_card_data.family_id, get_family_id(), FAMILY_ID_SIZE);
-  tap_card_data.retries = 5;
-
-  if (card_already_deleted_flash(flash_wallet_index, card_number)) {
-    if (card_number < 4)
-      flow_level.level_four++;
-    else {
-      flow_level.level_four = 1;
-      flow_level.level_three++;
+bool check_wallet_already_delelted_on_card(
+    card_delete_share_cfg_t *delete_config) {
+  uint8_t flash_wallet_index = 0xFF;
+  ASSERT(SUCCESS ==
+         get_index_by_name((const char *)delete_config->wallet->wallet_name,
+                           &flash_wallet_index));
+
+  if (card_already_deleted_flash(flash_wallet_index,
+                                 delete_config->card_number)) {
+    if (4 == delete_config->card_number) {
+      ASSERT(SUCCESS_ == delete_wallet_from_flash(flash_wallet_index));
     }
-    return;
+    return true;
   }
+  return false;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+card_error_type_e card_delete_share(card_delete_share_cfg_t *delete_config) {
+  card_operation_data_t card_data = {0};
+  card_error_type_e result = CARD_OPERATION_DEFAULT_INVALID;
+  char heading[50] = "";
+
+  if (true == check_wallet_already_delelted_on_card(delete_config)) {
+    return CARD_OPERATION_SUCCESS;
+  }
```

Comment:
Let's move this to card flow instead. So operation responsible to delete from the card when we know it is already deleted

Reply:
Will take up this change in the flow PR, added a TODO here.